src/context/abstractCast.ml

```diff
@@ -13,10 +13,17 @@ let make_static_call ctx c cf a pl args t p =
 			| [e] ->
 				let e,f = push_this ctx e in
 				ctx.with_type_stack <- (WithType.with_type t) :: ctx.with_type_stack;
+				let expected_type = apply_params a.a_params pl a.a_this in
 				let e = match ctx.g.do_macro ctx MExpr c.cl_path cf.cf_name [e] p with
-					| Some e -> type_expr ctx e WithType.value
+					| Some e -> type_expr ctx e (WithType.with_type expected_type)
 					| None ->  type_expr ctx (EConst (Ident "null"),p) WithType.value
 				in
+				let rec is_mono t = match t with
+					| TMono r -> (match !r with None -> true | Some t -> is_mono t)
+					| _ -> false
+				in
+				if not (is_mono e.etype) then
```

Comment:
Hmm, why this?

Could you add a test with an abstract over `Array<T>` so we end up with a situation where `t` _has_ a monomorph instead of being one?

Reply:
Because then monomorph will be bound to the underlying type. And if abstract has no `from UnderlyingType` then `var some:Asbtr = fromMacroValue` will fail here:
https://github.com/HaxeFoundation/haxe/blob/2985c9d8282003a35ade2971cbe0190300417e42/src/typing/typer.ml#L1445-L1446

Reply:
Your question made me realize, I chose wrong expected type to unify with. It should be an abstract itself, not the underlying type.

Reply:
After another round of thoughts I think that initial approach is correct.
We need to unify with the underlying type. But in addition we need to return a cast to that abstract.

Reply:
I still don't quite understand this. Why would the `@:from` method return the underlying type instead of the abstract? That's not how non-macro `@:from` methods operate either, so it seems to me that whatever the macro returns should be unified against the abstract type.

Reply:
I suppose by "unified against the abstract type" you're excluding further implicit casts? If not, this can get a bit messy ^^

Reply:
> I still don't quite understand this. Why would the `@:from` method return the underlying type instead of the abstract?

Because we already rely on that. E.g. here: 
https://github.com/HaxeFoundation/haxe/blob/096b447deeb0cf240b1cf18548b8af1cbba9c061/tests/unit/src/unit/issues/misc/Issue4306Macro.hx#L9-L13
And in the issue (#5880) @ncannasse definitely relies on that behavior too.

Reply:
Maybe so but I don't see how that makes it any more correct design-wise.

Reply:
I agree, that's not correct. But changing that is a breaking change.

Reply:
I'd consider it a bugfix...

Reply:
I'd be happy to change that to behave like other `@:from`s.
@ncannasse ?

Reply:
@Simn fixed that